File: experiments/problem1_nonperiodicity/analyze_period_two_minimizer_plateau_transport.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

namespace {
using u128 = unsigned __int128;
constexpr int MAXIMUM_COMPLEXITY = 25;
constexpr int SCHEDULE_CAP = 64;
constexpr std::array<int, 4> GAPS{2, 3, 4, 5};
constexpr std::uint64_t KNOWN_STATE = 0x1bcd3a7b3fdfbULL;

u128 forward_generator(char name, u128 state) {
    const u128 stepped = state ^ ((state << 1) | (state << 2));
    if (name == 't') return stepped;
    if (name == 'u') return stepped ^ 1;
    if (name == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::invalid_argument("unknown generator");
}

int bit_length(std::uint64_t value) {
    return value == 0 ? 0 : 64 - __builtin_clzll(value);
}

struct Schedule {
    std::uint64_t bits = 0;
    std::uint8_t length = 0;
};

Schedule forced_zero_schedule(std::uint64_t initial) {
    u128 state = initial;
    Schedule schedule;
    for (int step = 0; step < SCHEDULE_CAP; ++step) {
        const unsigned residue = static_cast<unsigned>(state & 15);
        char branch;
        if (residue == 7) branch = 'u';
        else if (residue == 11) branch = 't';
        else return schedule;
        const u128 tail = (state - 3) >> 2;
        state = forward_generator(branch, forward_generator('p', tail));
        schedule.bits = (schedule.bits << 1) | (branch == 'u' ? 1 : 0);
        ++schedule.length;
    }
    throw std::runtime_error("forced zero schedule reached cap");
}

struct PrefixKey {
    std::uint64_t bits;
    std::uint8_t length;
    bool operator==(const PrefixKey& other) const {
        return bits == other.bits && length == other.length;
    }
};

struct PrefixHash {
    std::size_t operator()(const PrefixKey& key) const noexcept {
        std::uint64_t value = key.bits ^ (std::uint64_t{key.length} << 57);
        value ^= value >> 30;
        value *= 0xbf58476d1ce4e5b9ULL;
        value ^= value >> 27;
        value *= 0x94d049bb133111ebULL;
        value ^= value >> 31;
        return static_cast<std::size_t>(value);
    }
};

PrefixKey prefix_key(const Schedule& schedule, int end) {
    return {
        schedule.bits >> (schedule.length - end),
        static_cast<std::uint8_t>(end),
    };
}

std::string decode(std::uint64_t bits, int length) {
    std::string word;
    word.reserve(length);
    for (int index = length - 1; index >= 0; --index) {
        word.push_back(((bits >> index) & 1) ? 'u' : 't');
    }
    return word;
}

bool admissible(const std::string& word) {
    return word.find("uu") == std::string::npos
        && word.find("ttttt") == std::string::npos
        && word.find("ututtu") == std::string::npos;
}

std::string extension(const std::vector<int>& gaps, bool include_final_u) {
    std::string word = "u";
    for (std::size_t index = 0; index < gaps.size(); ++index) {
        word += std::string(gaps[index] - 1, 't');
        if (index + 1 < gaps.size() || include_final_u) word += 'u';
    }
    return word;
}

bool schedule_has_at(
    const Schedule& schedule,
    int cut,
    const std::string& target
) {
    if (cut + static_cast<int>(target.size()) > schedule.length) return false;
    for (int index = 0; index < static_cast<int>(target.size()); ++index) {
        const int shift = schedule.length - 1 - (cut + index);
        const char observed = ((schedule.bits >> shift) & 1) ? 'u' : 't';
        if (observed != target[index]) return false;
    }
    return true;
}

struct Pattern {
    std::vector<int> gaps;
    std::string target;
    std::string complete;
};

void build_patterns_recursive(
    int return_count,
    std::vector<int>& current,
    std::vector<Pattern>& output
) {
    if (static_cast<int>(current.size()) == return_count) {
        const std::string target = extension(current, false);
        const std::string complete = extension(current, true);
        if (admissible(complete)) output.push_back({current, target, complete});
        return;
    }
    for (int gap : GAPS) {
        current.push_back(gap);
        build_patterns_recursive(return_count, current, output);
        current.pop_back();
    }
}

std::vector<Pattern> patterns(int return_count) {
    std::vector<Pattern> output;
    std::vector<int> current;
    build_patterns_recursive(return_count, current, output);
    return output;
}

struct Example {
    int complexity;
    std::uint64_t state;
    int cut;
    std::string base;
    std::vector<int> gaps;
    std::string schedule;
};

struct Summary {
    std::uint64_t outputs = 0;
    std::uint64_t eligible = 0;
    std::uint64_t two = 0;
    std::uint64_t three = 0;
    int maximum_schedule = 0;
    std::vector<Example> two_examples;
    std::vector<Example> three_examples;
};

Summary run_phase(
    char phase,
    const std::vector<Pattern>& two_patterns,
    const std::vector<Pattern>& three_patterns
) {
    std::unordered_set<std::uint64_t> states{phase == 'p' ? 3ULL : 1ULL};
    std::unordered_map<PrefixKey, std::uint8_t, PrefixHash> minimum;
    Summary summary;
    std::cout << "phase=" << phase << '\n';

    for (int complexity = 1; complexity <= MAXIMUM_COMPLEXITY; ++complexity) {
        std::vector<std::pair<std::uint64_t, Schedule>> rows;
        rows.reserve(states.size());
        const int expected = phase == 'p' ? 2 * complexity : 2 * complexity - 1;
        for (std::uint64_t state : states) {
            if (bit_length(state) != expected) {
                throw std::runtime_error("frontier bit-length law failed");
            }
            if ((state & 3) != 3) continue;
            const Schedule schedule = forced_zero_schedule(state);
            rows.emplace_back(state, schedule);
            summary.maximum_schedule = std::max(
                summary.maximum_schedule, static_cast<int>(schedule.length)
            );
            for (int end = 0; end <= schedule.length; ++end) {
                minimum.emplace(prefix_key(schedule, end), complexity);
            }
        }

        std::uint64_t level_two = 0;
        std::uint64_t level_three = 0;
        for (const auto& [state, schedule] : rows) {
            for (int cut = 0; cut <= schedule.length; ++cut) {
                const PrefixKey key = prefix_key(schedule, cut);
                if (minimum.at(key) != complexity) continue;
                const std::string base = decode(key.bits, cut);
                auto scan = [&] (
                    const std::vector<Pattern>& candidates,
                    std::uint64_t& level_count,
                    std::uint64_t& total_count,
                    std::vector<Example>& examples
                ) {
                    for (const Pattern& candidate : candidates) {
                        if (!schedule_has_at(schedule, cut, candidate.target)) continue;
                        if (!admissible(base + candidate.complete)) continue;
                        ++level_count;
                        ++total_count;
                        if (examples.size() < 4) {
                            examples.push_back({
                                complexity,
                                state,
                                cut,
                                base,
                                candidate.gaps,
                                decode(schedule.bits, schedule.length),
                            });
                        }
                    }
                };
                scan(
                    two_patterns,
                    level_two,
                    summary.two,
                    summary.two_examples
                );
                scan(
                    three_patterns,
                    level_three,
                    summary.three,
                    summary.three_examples
                );
            }
        }

        summary.outputs += states.size();
        summary.eligible += rows.size();
        if (level_two || level_three) {
            std::cout << "candidate_level=" << complexity
                      << " two=" << level_two
                      << " three=" << level_three << '\n';
        }

        if (complexity != MAXIMUM_COMPLEXITY) {
            std::unordered_set<std::uint64_t> next;
            next.reserve(states.size() * 3);
            for (std::uint64_t state : states) {
                const std::uint64_t stepped =
                    state ^ ((state << 1) | (state << 2));
                next.insert(stepped);
                next.insert(stepped ^ 1);
                if ((state & 1) == 0) next.insert(stepped ^ 3);
            }
            states.swap(next);
        }
    }

    std::cout << "phase_outputs_checked=" << summary.outputs << '\n';
    std::cout << "phase_eligible_outputs=" << summary.eligible << '\n';
    std::cout << "phase_max_schedule=" << summary.maximum_schedule << '\n';
    std::cout << "phase_two_return_candidates=" << summary.two << '\n';
    std::cout << "phase_three_return_candidates=" << summary.three << '\n';
    for (const Example& example : summary.two_examples) {
        std::cout << "two_example complexity=" << example.complexity
                  << " state=0x" << std::hex << example.state << std::dec
                  << " cut=" << example.cut
                  << " base=" << example.base
                  << " gaps=";
        for (std::size_t index = 0; index < example.gaps.size(); ++index) {
            if (index) std::cout << ',';
            std::cout << example.gaps[index];
        }
        std::cout << " schedule=" << example.schedule << '\n';
    }
    return summary;
}
}  // namespace

int main() {
    const std::vector<Pattern> two_patterns = patterns(2);
    const std::vector<Pattern> three_patterns = patterns(3);
    std::cout << "maximum_complexity=" << MAXIMUM_COMPLEXITY << '\n';
    std::cout << "two_patterns=" << two_patterns.size() << '\n';
    std::cout << "three_patterns=" << three_patterns.size() << '\n';

    const Summary p = run_phase('p', two_patterns, three_patterns);
    const Summary u = run_phase('u', two_patterns, three_patterns);
    std::cout << "total_outputs_checked=" << p.outputs + u.outputs << '\n';
    std::cout << "total_eligible_outputs=" << p.eligible + u.eligible << '\n';
    std::cout << "total_two_return_candidates=" << p.two + u.two << '\n';
    std::cout << "total_three_return_candidates=" << p.three + u.three << '\n';

    if (p.two != 0 || u.two != 1 || p.three != 0 || u.three != 0) {
        throw std::runtime_error("unexpected plateau-candidate totals");
    }
    if (u.two_examples.empty() || u.two_examples.front().state != KNOWN_STATE) {
        throw std::runtime_error("known complexity-25 candidate was not recovered");
    }

    const Schedule schedule = forced_zero_schedule(KNOWN_STATE);
    const std::string word = decode(schedule.bits, schedule.length);
    std::cout << "known_forced_schedule=" << word << '\n';
    if (word != "tutututttutututt") {
        throw std::runtime_error("known forced schedule changed");
    }
    std::cout << "known_third_required_branch=u\n";
    std::cout << "known_observed_branch=t\n";
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_minimizer_plateau_transport.py
#!/usr/bin/env python3
"""Exact fixed-complexity plateau transport for period-two phase minimizers.

For phase a, complexity k, depth L and residue X modulo 4**L, let
C_(a,k)(X,L) be the exact phase-frontier cylinder. If X' extends X by an
r-digit base-four block c, then the same-complexity cylinder is obtained by
retaining exactly those residual states whose next r digits equal c and then
shifting the consumed block away. This gives an exact return-plateau survival
criterion without searching any higher complexity.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from functools import lru_cache
from itertools import product
from typing import Any

PHASES = ("p", "u")
GAPS = (2, 3, 4, 5)
FORBIDDEN = ("uu", "ttttt", "ututtu")
CHILD_DIGIT_MASK = {0: 0b1011, 1: 0b1100, 2: 0b1110, 3: 0b0011}
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
DEFAULT_SCHEDULE_CAP = 64
KNOWN = 0x1BCD3A7B3FDFB
KNOWN_BASE = "tutututttutu"
KNOWN_MIDDLE = KNOWN_BASE + "tu"
KNOWN_FINAL = KNOWN_MIDDLE + "tu"
EXPECTED = {
    1: {"p": 1, "u": 2},
    5: {"p": 8, "u": 12},
    7: {"p": 13, "u": 14},
    12: {"p": 28, "u": 27},
    14: {"p": 33, "u": 30},
}


class CampaignLimitError(RuntimeError):
    """Raised before a controlled campaign exceeds its configured cap."""


def forward_generator(name: str, state: int) -> int:
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    stepped = forward_generator("t", state)
    if state & 1:
        return stepped, stepped ^ 1
    return stepped, stepped ^ 1, stepped ^ 3


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def expected_bits(phase: str, complexity: int) -> int:
    return 2 * complexity if phase == "p" else 2 * complexity - 1


def inverse_t(output: int) -> int | None:
    if output == 0:
        return 0
    width = output.bit_length() - 2
    if width <= 0:
        return None
    state = 0
    for position in range(width):
        lower = 0
        if position >= 1:
            lower |= (state >> (position - 1)) & 1
        if position >= 2:
            lower |= (state >> (position - 2)) & 1
        state |= ((((output >> position) & 1) ^ lower) << position)
    return state if forward_generator("t", state) == output else None


def inverse_generator(name: str, output: int) -> int | None:
    if name == "t":
        state = inverse_t(output)
    elif name == "u":
        state = inverse_t(output ^ 1)
    elif name == "p":
        recovered_low_bit = (output & 1) ^ 1
        state = inverse_t(output ^ 1 ^ (2 if recovered_low_bit == 0 else 0))
    else:
        raise ValueError("unknown generator")
    if state is None or forward_generator(name, state) != output:
        return None
    return state


def candidate_parent(quotient: int, parent_digit: int) -> int | None:
    name = "t" if parent_digit == 0 else "u" if parent_digit == 1 else "p"
    residual = inverse_generator(name, quotient)
    return None if residual is None else 4 * residual + parent_digit


def build_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {phase_start(phase)}]
    for _ in range(2, maximum_complexity + 1):
        levels.append(
            {child for state in levels[-1] for child in frontier_children(state)}
        )
    return levels


def make_member(levels: dict[str, list[set[int]]], base_cap: int):
    @lru_cache(maxsize=None)
    def member(phase: str, complexity: int, target: int) -> bool:
        if complexity < 1 or target < 0:
            return False
        if target.bit_length() != expected_bits(phase, complexity):
            return False
        if complexity <= base_cap:
            return target in levels[phase][complexity]
        quotient, digit = divmod(target, 4)
        for parent_digit in range(4):
            if not (CHILD_DIGIT_MASK[parent_digit] >> digit) & 1:
                continue
            parent = candidate_parent(quotient, parent_digit)
            if parent is not None and member(phase, complexity - 1, parent):
                return True
        return False

    return member


def inverse_t_mod(output: int, width: int) -> int:
    output &= (1 << width) - 1 if width else 0
    state = 0
    for position in range(width):
        lower = 0
        if position >= 1:
            lower |= (state >> (position - 1)) & 1
        if position >= 2:
            lower |= (state >> (position - 2)) & 1
        state |= ((((output >> position) & 1) ^ lower) << position)
    return state


def inverse_generator_mod(name: str, output: int, width: int) -> int:
    output &= (1 << width) - 1 if width else 0
    if name == "t":
        return inverse_t_mod(output, width)
    if name == "u":
        return inverse_t_mod(output ^ 1, width)
    if name == "p":
        recovered_low_bit = (output & 1) ^ 1
        return inverse_t_mod(
            output ^ 1 ^ (2 if recovered_low_bit == 0 else 0), width
        )
    raise ValueError("unknown generator")


def survivor_for_word(word: str) -> int:
    state = 0
    precision = 0
    for branch in reversed(word):
        precision += 2
        inner_width = precision - 2
        state = inverse_generator_mod(branch, state, inner_width)
        state = inverse_generator_mod("p", state, inner_width)
        state = ((state << 2) | 3) & ((1 << precision) - 1)
    return state


def fringe_step(state: int) -> int:
    row = 1 + 2 * state
    odd = row ^ ((row >> 1) | (row >> 2))
    return (odd << 1) ^ (odd | (odd >> 1))


def actual_driver(length: int) -> str:
    state = 0
    word: list[str] = []
    for _ in range(length):
        word.append("u" if state & 3 == 0 else "t")
        state = fringe_step(state)
    return "".join(word)


def forced_zero_schedule(state: int, cap: int = DEFAULT_SCHEDULE_CAP) -> str:
    word: list[str] = []
    for _ in range(cap):
        residue = state & 15
        if residue == 7:
            branch = "u"
        elif residue == 11:
            branch = "t"
        else:
            return "".join(word)
        state = forward_generator(
            branch, forward_generator("p", (state - 3) >> 2)
        )
        word.append(branch)
    raise CampaignLimitError("forced zero schedule reached the safety cap")


def admissible(word: str) -> bool:
    return not any(factor in word for factor in FORBIDDEN)


def return_extension(gaps: tuple[int, ...], include_final_u: bool) -> str:
    word = "u"
    for index, gap in enumerate(gaps):
        word += "t" * (gap - 1)
        if index < len(gaps) - 1 or include_final_u:
            word += "u"
    return word


def cylinder_filter(
    phase: str,
    depth: int,
    complexity: int,
    residue: int,
    levels: dict[str, list[set[int]]],
    member,
) -> list[int]:
    member.cache_clear()
    if complexity <= depth:
        return [residue] if member(phase, complexity, residue) else []
    residual_complexity = complexity - depth
    current = sorted(levels[phase][residual_complexity])
    for index in range(depth - 1, -1, -1):
        digit = (residue >> (2 * index)) & 3
        next_complexity = residual_complexity + (depth - index)
        current = [
            4 * quotient + digit
            for quotient in current
            if member(phase, next_complexity, 4 * quotient + digit)
        ]
        if not current:
            break
    return current


def plateau_transport(
    states: list[int],
    old_depth: int,
    new_depth: int,
    old_residue: int,
    new_residue: int,
) -> dict[str, Any]:
    if new_depth <= old_depth:
        raise ValueError("new depth must be larger")
    if new_residue % (4**old_depth) != old_residue:
        raise ValueError("cylinders are not nested")
    span = new_depth - old_depth
    block = (new_residue - old_residue) // (4**old_depth)
    mask = 4**span - 1
    survivors = [
        state
        for state in states
        if ((state >> (2 * old_depth)) & mask) == block
    ]
    old_residuals = [state >> (2 * old_depth) for state in states]
    shifted = [state >> (2 * new_depth) for state in survivors]
    predicted = [
        residual >> (2 * span)
        for residual in old_residuals
        if residual & mask == block
    ]
    if shifted != predicted:
        raise AssertionError("residual transport identity failed")
    return {
        "span": span,
        "block": block,
        "block_base4_digits": [(block >> (2 * i)) & 3 for i in range(span)],
        "input_count": len(states),
        "survivor_count": len(survivors),
        "survivor_states_hex": [hex(state) for state in survivors],
        "shifted_residuals_hex": [hex(state) for state in shifted],
    }


def exhaustive_transport(levels: dict[str, list[set[int]]]) -> int:
    checks = 0
    for phase in PHASES:
        for complexity in range(1, 9):
            full = sorted(levels[phase][complexity])
            for new_depth in range(2, min(5, complexity + 1) + 1):
                for old_depth in range(1, new_depth):
                    old_modulus = 4**old_depth
                    new_modulus = 4**new_depth
                    groups: dict[int, list[int]] = {}
                    for state in full:
                        groups.setdefault(state % old_modulus, []).append(state)
                    for new_residue in {state % new_modulus for state in full}:
                        old_residue = new_residue % old_modulus
                        row = plateau_transport(
                            groups[old_residue],
                            old_depth,
                            new_depth,
                            old_residue,
                            new_residue,
                        )
                        direct = [
                            state
                            for state in groups[old_residue]
                            if state % new_modulus == new_residue
                        ]
                        if row["survivor_states_hex"] != [hex(state) for state in direct]:
                            raise AssertionError("transport disagreed with direct grouping")
                        checks += 1
    return checks


def frontier_plateau_census(
    maximum_complexity: int, schedule_cap: int
) -> dict[str, Any]:
    patterns = {
        return_count: [
            (
                gaps,
                return_extension(gaps, False),
                return_extension(gaps, True),
            )
            for gaps in product(GAPS, repeat=return_count)
            if admissible(return_extension(gaps, True))
        ]
        for return_count in (2, 3)
    }
    result: dict[str, Any] = {}
    for phase in PHASES:
        states = {phase_start(phase)}
        minimum: dict[str, int] = {}
        totals = {2: 0, 3: 0}
        examples: dict[int, list[dict[str, Any]]] = {2: [], 3: []}
        outputs = 0
        eligible = 0
        candidate_levels: list[dict[str, int]] = []
        for complexity in range(1, maximum_complexity + 1):
            rows: list[tuple[int, str]] = []
            for state in states:
                if state & 3 != 3:
                    continue
                schedule = forced_zero_schedule(state, schedule_cap)
                rows.append((state, schedule))
                eligible += 1
                for end in range(len(schedule) + 1):
                    minimum.setdefault(schedule[:end], complexity)
            level_counts = {2: 0, 3: 0}
            for state, schedule in rows:
                for cut in range(len(schedule) + 1):
                    prefix = schedule[:cut]
                    if minimum[prefix] != complexity:
                        continue
                    for return_count in (2, 3):
                        for gaps, target, complete in patterns[return_count]:
                            if not schedule[cut:].startswith(target):
                                continue
                            if not admissible(prefix + complete):
                                continue
                            totals[return_count] += 1
                            level_counts[return_count] += 1
                            if len(examples[return_count]) < 4:
                                examples[return_count].append(
                                    {
                                        "complexity": complexity,
                                        "state_hex": hex(state),
                                        "cut": cut,
                                        "base_prefix": prefix,
                                        "gaps": list(gaps),
                                        "forced_schedule": schedule,
                                    }
                                )
            if level_counts[2] or level_counts[3]:
                candidate_levels.append(
                    {
                        "complexity": complexity,
                        "two_return": level_counts[2],
                        "three_return": level_counts[3],
                    }
                )
            outputs += len(states)
            states = {
                child for state in states for child in frontier_children(state)
            }
        result[phase] = {
            "outputs_checked": outputs,
            "eligible_outputs": eligible,
            "two_return_candidates": totals[2],
            "three_return_candidates": totals[3],
            "candidate_levels": candidate_levels,
            "examples": examples,
        }
    return result


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
    schedule_cap: int = DEFAULT_SCHEDULE_CAP,
) -> dict[str, Any]:
    if not 8 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise CampaignLimitError("maximum complexity outside controlled range")
    levels = {
        phase: build_levels(phase, maximum_complexity) for phase in PHASES
    }
    member = make_member(levels, maximum_complexity)
    transport_checks = exhaustive_transport(levels)

    driver = actual_driver(20)
    return_depths = [
        index + 1
        for index, branch in enumerate(driver)
        if branch == "u" and index + 1 <= 14
    ]
    actual_rows: list[dict[str, Any]] = []
    for old_depth, new_depth in zip(return_depths, return_depths[1:]):
        old_residue = survivor_for_word(driver[:old_depth])
        new_residue = survivor_for_word(driver[:new_depth])
        phase_rows: dict[str, Any] = {}
        for phase in PHASES:
            complexity = EXPECTED[old_depth][phase]
            states = cylinder_filter(
                phase,
                old_depth,
                complexity,
                old_residue,
                levels,
                member,
            )
            phase_rows[phase] = {
                "complexity": complexity,
                "minimum_count": len(states),
                "transport": plateau_transport(
                    states,
                    old_depth,
                    new_depth,
                    old_residue,
                    new_residue,
                ),
            }
        actual_rows.append(
            {
                "old_depth": old_depth,
                "new_depth": new_depth,
                "gap": new_depth - old_depth,
                "phases": phase_rows,
            }
        )

    known_schedule = forced_zero_schedule(KNOWN)
    known_transports = []
    for old_word, new_word in (
        (KNOWN_BASE, KNOWN_MIDDLE),
        (KNOWN_MIDDLE, KNOWN_FINAL),
    ):
        known_transports.append(
            plateau_transport(
                [KNOWN],
                len(old_word),
                len(new_word),
                survivor_for_word(old_word),
                survivor_for_word(new_word),
            )
        )
    if known_schedule.startswith(KNOWN_FINAL):
        raise AssertionError("known state unexpectedly forced the final branch")

    payload: dict[str, Any] = {
        "status": "exact-plateau-transport-and-finite-census",
        "maximum_complexity": maximum_complexity,
        "transport_theorem": (
            "For nested depth-L and depth-(L+r) cylinders at fixed complexity, "
            "retain precisely residual minimizers congruent to the extension "
            "block modulo 4^r, then divide their residuals by 4^r. Zero penalty "
            "is equivalent to a nonempty filtered set."
        ),
        "exhaustive_small_transport_checks": transport_checks,
        "actual_return_transports": actual_rows,
        "known_counterexample": {
            "state_hex": hex(KNOWN),
            "forced_schedule": known_schedule,
            "two_zero_penalty_transports": known_transports,
            "third_return_possible": False,
            "reason": (
                "the required u at the next return is the depth-16 branch, which "
                "is invisible modulo 4^16; the state forces t there"
            ),
        },
        "frontier_census": frontier_plateau_census(
            maximum_complexity, schedule_cap
        ),
        "scientific_boundary": (
            "The transport theorem is all-depth exact. The absence of three-return "
            "candidates is finite through the configured complexity and does not "
            "prove the all-depth isolated-three-return lemma, phase-complexity "
            "divergence, exclusion of period two, or Rule 30 center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity",
        type=int,
        default=DEFAULT_MAXIMUM_COMPLEXITY,
    )
    parser.add_argument("--schedule-cap", type=int, default=DEFAULT_SCHEDULE_CAP)
    args = parser.parse_args()
    print(
        json.dumps(
            run_campaign(args.maximum_complexity, args.schedule_cap),
            indent=2,
            sort_keys=True,
        )
    )


if __name__ == "__main__":
    main()
